src/shared/tpm2-util.c

```diff
@@ -458,24 +458,161 @@ static int tpm2_make_primary(
+
+#define FOREACH_PCR_IN_TPMS_PCR_SELECTION(pcr, tpms)                    \
+        FOREACH_PCR_IN_MASK(pcr, tpm2_tpms_pcr_selection_to_mask(tpms))
+
+#define FOREACH_TPMS_PCR_SELECTION_IN_TPML_PCR_SELECTION(tpms, tpml)    \
+        for (TPMS_PCR_SELECTION *tpms = (tpml)->pcrSelections;          \
+             (uint32_t)(tpms - (tpml)->pcrSelections) < (tpml)->count;  \
+             tpms++)
+
+#define FOREACH_PCR_IN_TPML_PCR_SELECTION(pcr, tpms, tpml)              \
+        FOREACH_TPMS_PCR_SELECTION_IN_TPML_PCR_SELECTION(tpms, tpml)    \
+                FOREACH_PCR_IN_TPMS_PCR_SELECTION(pcr, tpms)            \
+
+static char *tpm2_tpms_pcr_selection_to_string(const TPMS_PCR_SELECTION *s) {
+        const char *algstr = strna(tpm2_hash_alg_to_string(s->hash));
+
+        _cleanup_free_ char *maskstr = tpm2_pcr_mask_to_string(tpm2_tpms_pcr_selection_to_mask(s));
+        if (!maskstr)
+                return NULL;
+
+        return strjoin(algstr, "(", maskstr, ")");
+}
+
+/* Utility functions for TPML_PCR_SELECTION. */
+
+static TPMS_PCR_SELECTION *tpm2_tpml_pcr_selection_get_tpms_pcr_selection(TPML_PCR_SELECTION *l, TPMI_ALG_HASH hash) {
+        FOREACH_TPMS_PCR_SELECTION_IN_TPML_PCR_SELECTION(s, l)
+                if (s->hash == hash)
+                        return s;
+        return NULL;
+}
+
+uint32_t tpm2_tpml_pcr_selection_to_mask(const TPML_PCR_SELECTION *l, TPMI_ALG_HASH hash) {
+        TPMS_PCR_SELECTION *s = tpm2_tpml_pcr_selection_get_tpms_pcr_selection((TPML_PCR_SELECTION*) l, hash);
+        return s ? tpm2_tpms_pcr_selection_to_mask(s) : 0;
+}
+
+TPML_PCR_SELECTION tpm2_tpml_pcr_selection_from_mask(uint32_t mask, TPMI_ALG_HASH hash) {
+        return (TPML_PCR_SELECTION) {
                 .count = 1,
-                .pcrSelections[0] = {
-                        .hash = bank,
-                        .sizeofSelect = 3,
-                        .pcrSelect[0] = mask & 0xFF,
-                        .pcrSelect[1] = (mask >> 8) & 0xFF,
-                        .pcrSelect[2] = (mask >> 16) & 0xFF,
-                }
+                .pcrSelections[0] = tpm2_tpms_pcr_selection_from_mask(mask, hash),
         };
 }
 
+static unsigned tpm2_tpml_pcr_selection_weight(const TPML_PCR_SELECTION *l) {
+        unsigned weight = 0;
+        for (unsigned i = 0; i < l->count; i++)
+                weight += tpm2_tpms_pcr_selection_weight(&l->pcrSelections[i]);
+        return weight;
+}
+
+#define tpm2_tpml_pcr_selection_empty(l) (tpm2_tpml_pcr_selection_weight(l) == 0)
+
+static void tpm2_tpml_pcr_selection_add_tpms_pcr_selection(TPML_PCR_SELECTION *l, const TPMS_PCR_SELECTION *s) {
+        assert(l->count < TPM2_NUM_PCR_BANKS);
+        l->pcrSelections[l->count++] = *s;
+}
+
+/* This verifies all pcrSelection[] entries have a unique hash, and combines any duplicates. */
+static void tpm2_tpml_pcr_selection_normalize(TPML_PCR_SELECTION *l) {
+        TPML_PCR_SELECTION newl = { .count = 0, };
```

Comment:
just write `{}` which will initialize the whole struct to all zeroes. (this is a gcc extension that doesn't exist in ANSI C)

Reply:
I reworked this bit, to roll it into other functions.